mopidy/audio/scan.py

```diff
@@ -51,68 +40,117 @@ def scan(self, uri):
-        self._bus.set_flushing(True)
-        self._pipe.set_state(gst.STATE_NULL)
-
-    def _query_duration(self):
-        try:
-            duration = self._pipe.query_duration(gst.FORMAT_TIME, None)[0]
-        except gst.QueryError:
-            return None
-
-        if duration < 0:
-            return None
-        else:
-            return duration // gst.MSECOND
+            pipeline.set_state(gst.STATE_NULL)
+            del pipeline
+
+        return _Result(uri, tags, duration, seekable, mime)
+
+
+# Turns out it's _much_ faster to just create a new pipeline for every as
+# decodebins and other elements don't seem to take well to being reused.
+def _setup_pipeline(uri, proxy_config=None):
+    src = gst.element_make_from_uri(gst.URI_SRC, uri)
+    if not src:
+        raise exceptions.ScannerError('GStreamer can not open: %s' % uri)
+
+    typefind = gst.element_factory_make('typefind')
+    decodebin = gst.element_factory_make('decodebin2')
+    sink = gst.element_factory_make('fakesink')
+
+    pipeline = gst.element_factory_make('pipeline')
+    pipeline.add_many(src, typefind, decodebin, sink)
+    gst.element_link_many(src, typefind, decodebin)
+
+    if proxy_config:
+        utils.setup_proxy(src, proxy_config)
+
+    decodebin.set_property('caps', _RAW_AUDIO)
+    decodebin.connect('pad-added', _pad_added, sink)
+    typefind.connect('have-type', _have_type, decodebin)
+
+    return pipeline
+
+
+def _have_type(element, probability, caps, decodebin):
+    decodebin.set_property('sink-caps', caps)
+    msg = gst.message_new_application(element, caps.get_structure(0))
+    element.get_bus().post(msg)
+
+
+def _pad_added(element, pad, sink):
+    return pad.link(sink.get_pad('sink'))
+
+
+def _start_pipeline(pipeline):
+    if pipeline.set_state(gst.STATE_PAUSED) == gst.STATE_CHANGE_NO_PREROLL:
+        pipeline.set_state(gst.STATE_PLAYING)
+
+
+def _query_duration(pipeline):
+    try:
+        duration = pipeline.query_duration(gst.FORMAT_TIME, None)[0]
+    except gst.QueryError:
+        return None
+
+    if duration < 0:
+        return None
+    else:
+        return duration // gst.MSECOND
+
+
+def _query_seekable(pipeline):
+    query = gst.query_new_seeking(gst.FORMAT_TIME)
+    pipeline.query(query)
+    return query.parse_seeking()[1]
+
+
+def _process(pipeline, timeout):
+    start = time.time()
+    tags, mime, missing_description = {}, None, None
+    bus = pipeline.get_bus()
+
+    while time.time() - start < timeout:
+        if not bus.have_pending():
+            continue
+        message = bus.pop()
+
+        if message.type == gst.MESSAGE_ELEMENT:
+            if gst.pbutils.is_missing_plugin_message(message):
+                missing_description = encoding.locale_decode(
+                    _missing_plugin_desc(message))
+        elif message.type == gst.MESSAGE_APPLICATION:
+            mime = message.structure.get_name()
+            if mime.startswith('text/') or mime == 'application/xml':
```

Comment:
From the `decodebin2` code I know that the only thing the explicitly mess with, and break for us is `text/plain` other things I might be able to let through and handle in a more generic way. Waiting for the error and then using the MIME, though that being said that could probably also be done for the `text/plain` case.

So the contract would basically be that we `raise ScannerError(message, mime=mime)` and then you can do stuff based on that instead of having MIME in the result.